[Shine] 스프링 카페 6단계 - 댓글

zbqmgldjfh opened the pull request:
# 안녕하세요! Shine 입니다!
우선 리뷰를 해주시는 리뷰어 님께 감사의 인사를 전합니다!

## 이번 과제를 수행하며...
요구사항에서는 게시물에 자신이 쓴 댓글이 있으면 삭제 불가라 되어있지만...
그냥 CASCADE 옵션을 주어 본문 삭제시 관련 댓글까지 모두 삭제돠도록 처리하였습니다.

## 질문
이번 댓글 기능 구현 에 사용한 DTO는 다음과 같습니다.
```java
public class ReplyDto {
    private String userId;
    private String contents;
    // 생략...
}
```
이를 통해 사용자로부터 댓글의 내용과, 세션에 저장된 userId를 넘겨 받게 됩니다.
여기서 질문이 있습니다!

1. 뷰에서 session정보 함께 넘기기
userId를 html상에서 제출을 누를때 Thymeleaf의 session 속성을 통해 contents와 함께 보내고 있습니다.
왜냐하면 사용자가 댓글을 달기 위해 넘어오는 정보에 contents와 session 정보가 담겨있어야 한다고 생각했기 때문입니다.
```html
<input type="hidden" th:value="${session.SESSIONED_USER.userId}" name="userId">
```

2. 컨트롤러에서 처리하기
아니면 그냥 Controller에서 contents만 받은후, 컨트롤러에서 session에 접근하여 userId를 가지고 오는 방법이 있습니다.

일단 1번 방식을 선택했습니다. 어떤 방식이 더 적합할까요? =>  다시 일단 2번 방식으로 변경 (생각이 왔다 갔다...)

## Step06 ToDO-List

- [x] Reply Class 만들기
- [x] Reply JDBC 로 연결
- [x] 답변 달기 기능 추가 (로그인 한 사용자만)
- [x] 답변 목록 보여주기
- [x] 자신이 쓴 댓글 삭제 가능 (이번에는 히든필드를 통해 delete 메서드 이용)
  추가 요구사항은 미 반영

## Step05 리뷰 내용 수정 사항
- [x] Save, Delete 시 Long 값을 반환하게 했던 이유 고민해보기
- [x] SESSIONED_USER 는 상수로 관리하기
- [x] sessioned_user 변수명 camelCase 로 변경 (변경을 좀 하자면요ㅠ,ㅠ 인텔리제이 자동완성...)
- [ ] PUT 방식으로 변경해 보기 (아직 변경 못함)

## 실행 화면
<img width="1102" alt="스크린샷 2022-03-17 오후 4 21 33" src="https://user-images.githubusercontent.com/60593969/158757487-f7d773aa-f894-4191-9e5e-c65169203ba4.png">

zbqmgldjfh committed:
docs(README): Step06 Todo List 작성 완료

zbqmgldjfh committed:
feat(DbReplyRepository): 댓글 작성을 위한 DbReplyRepository 구현하기

zbqmgldjfh committed:
docs(README): Step05 수정 사항 todo list 작성

zbqmgldjfh committed:
feat(SessionConst): 전체에서 공통적으로 사용하는 상수 분리

- 변경 전
이전에는 코드 전체에서 각 class 별로 상수를 만들어 사용하였다.

- 변경 후
아예 SessionConst를 만들어 사용하게 되었다.

zbqmgldjfh committed:
style: 변수명을 sessioned_user 에서 sessionedUser로 변경

zbqmgldjfh committed:
feat(schema.sql): Reply에 제약조건 이름 reply_pk 추가해주기

zbqmgldjfh committed:
feat(ReplyController): Reply 추가 기능과, 목록 보여주기 기능 구현 완료

zbqmgldjfh committed:
chore: 주석 처리된 부분 제거

zbqmgldjfh committed:
feat(ReplyController): 댓글 삭제 기능 구현 완료

zbqmgldjfh committed:
test(ArticleControllerTest): ArticleController에 추가된 replyService를 mock으로 추가해주기

zbqmgldjfh committed:
docs(README): 구현 사항 반영

zbqmgldjfh committed:
setting: 스프링 설정을 통해 매번 실행시 데이터 초기화 되도록 변경

zbqmgldjfh committed:
setting: system 설정 파일로 java.version 이동

해로쿠 패포를 위한 11버전 설정

zbqmgldjfh committed:
setting: plain jar 생성 방지용

zbqmgldjfh committed:
setting: System 으로 디렉토리 이동

zbqmgldjfh committed:
setting: System 으로 디렉토리 이동2

zbqmgldjfh committed:
setting: WebConfig 에 "/*.ico" 제외시키기

zbqmgldjfh committed:
setting: H2 database를 in-memory로 변경

zbqmgldjfh committed:
setting: ExcluePattern에 "/error" 추가

wheejuni commented on src/main/java/com/kakao/cafe/repository/Query.java:
```diff
@@ -14,4 +14,10 @@
     DELETE_ARTICLE,
+
+    // reply
+    SAVE_REPLY,
+    SELECT_REPLY,
+    SELECT_REPLYS,
+    DELETE_REPLY,
```
이런 방식은 조금 피해달라고 리뷰드린 적 있는 것 같습니다^^
쿼리를 이렇게 열거형으로 관리해서 얻는 이점이 무엇인지요? 궁금합니다.

zbqmgldjfh replied:
음?? 아마 저말고 다른분께 남긴 리뷰이신것 같습니다 ㅎㅎ
혹시나 하는 마음에 저의 이전에 리뷰해주신 2단계리뷰를 다시 살펴봤는데 처음 말씀해주신 리뷰 입니다 ㅎㅎ!!

쿼리를 열거형으로 만들었을때 : 
Query 만 처도 Query. ~~~ 가 목록으로 보여지니 이는 인텔리제이를 사용하는 입장에서 장점이 될수는없을까요? ㅠ,ㅠ
Query에 해당되는 이미 내가 구현한 목록을 확인할수 있으니 이는 장점이 되지 않을까요??
일종의 이용 가능한 쿼리 명세표 같은것 이죠! 만약 하드 코딩으로 각각의 코드에 있다면 프로젝트 전체에서 내가 어떤 쿼리들을 사용하고 있는지 한눈에 이해하기 힘들것 같습니다?

wheejuni commented on src/main/java/com/kakao/cafe/repository/db/DbReplyRepository.java:
```diff
@@ -0,0 +1,88 @@
+    @Override
+    public Long save(Reply reply) {
+        DbTemplate template = new DbTemplate(dataSource);
+        String SQL = queryLoader.get(Query.SAVE_REPLY);
```
제가 흘끗 보긴 했습니다만 `Query` 에도 쿼리는 없던데요.
아, `QueryLoader` 가 있군요. 그러면 새로 같이 개발하는 개발자는 `Query` 를 확인한 다음에 `QueryLoader` 를 타고 들어가서 작동 원리를 파악한 다음 쿼리가 작성된 정적 파일을 새로 열어서 쿼리를 수정하면 되는 걸까요. 왜 그래야 하는거죠.

쿼리가 여기에 있으면 바로 수정할 수 있을 텐데요. 상기한 방식이 정말 좋은 개발이라고 생각하시는지 궁금합니다^^

wheejuni replied:
Java 언어에서 멀티라인 스트링이 지원되지 않아 지저분하다고 생각하실 수도 있는데 이걸 IDE의 기능으로 풀어낼 수도 있습니다. 한번 찾아보시기 바랍니다.

zbqmgldjfh replied:
우선 쿼리는 리소스의 query.yml 파일에 모아있습니다!

Query 로더를 사용하지 않던 기존 방식
JDBC의 각각의 파일에 직접적으로 SQL문이 공존한다. 따라서 해당 코드를 읽기에 불편함은 없다.
다만 관리면에서 복잡해 진다. repository가 겨우 3개 뿐인대도 점점 SQL문 관리가 너무 분산되기 시작했다.

Query 로더를 사용한후 저의 생각
query.yml에 모든 쿼리가 다 모여 있다. 특정 쿼리의 조건은 바꿔줘야 할때 해당 파일을 찾아가는 것 이 아니라, 단순히 query.yml문만 열면 끝이다. 이곳에 전부 쿼리가 모여있기 때문이다!

새롭게 같이 하는 개발자 입장에서도 사용하는 모든 SQL문은 query.yml 문에서 확인이 가능하기 때문에 한눈에 전체 SQL문을 이해할수 있다.
또한 사용하는 부분에서는 단지 loader(Query.XXX) 와 같은 "아~ loader에서 로딩해오는 구나" 라고 생각하고 사용하면 끝이다?

(사실 여러명이 다루게 된다는 미래의 이야기는 생각 안해봤는데 이번 기회에 해보게 되었습니다.
오히려 여러명이 함께 사용하니 query.yml에 모든 쿼리를 배치하고, loader를 통해 가져다 사용하는것이 더 좋지 않나요??
**MyBatis도 사실상 외부에서 xml 로 쿼리를 관리하고 필요할때 로딩해 사용하면서 유지보수성을 높힌것 아닌가요?**)

음 리뷰어님의 의견에 단순 순응하기 보다는 좀더 의견을 듣고싶습니다!
이러한 queryLoader를 사용하는방식이 진짜 장점이라 생각하는데, 잘못된것 일까요 ㅠㅡㅠ

zbqmgldjfh replied:
java 언어에서 멀티라인 스트링 은 처음 듣는 키워드 라 공부해보겠습니다!!
공부 키워드를 알려주셔서 감사합니다!!

wheejuni replied:
마이바티스가 유지보수성이 높다고 생각하지 않습니다^^ 여기서부터 생각이 다른 것 같군요.
마이바티스를 쓸 떄마다 전 욕하면서 사용했었거든요. 이 관점이 다르다면 제 의견을 수용하지 않으셔도 좋습니다. ^^

wheejuni replied:
코드를 작성할 때에 단순 효율성도 중요하지만 _사람이 하는 일이다_ 라는 관점도 대단히 중요하다고 생각하는데요,
쿼리와 로직이 멀리 떨어져 있으면 있을수록 특정 컴포넌트에 경도되는 개발 플로우가 형성될 거란 우려도 있습니다. 즉 지금 구현하신 방향이나 마이바티스처럼 두 요소를 멀리 떨어트려놓으면, 효율성은 오를지도(?) 모르겠습니다만, **결국 누군가는 쿼리만 작성하고, 누군가는 로직만 짜는 불상사** 가 발생하기 그만큼 쉬워진다는 생각도 듭니다. 실제로 아직도 많은 회사들이 그렇게 일하고 있고요.

그런 방향이 바람직한가? 는 각자 생각이 다를 것 같기는 합니다. ^^

zbqmgldjfh replied:
감사합니다 덕분에 조금더 시야를 넓게 생각해보는 리뷰였던것 같습니다 ㅎㅎ
리뷰어 님의 의견을 공유하고 저의 생각을 전달할 수 있어서 진심으로 최근 받은 리뷰중 가장 좋은 리뷰 인것 같습니다.

아마 MyBatis처럼 외부 관리가 편하다 한 저의 이유는....
이전까지 혼자서만 개발하다보니 그런점이 큰 것 같습니다 ㅠ.ㅠ... 
향후 팀원들과 협업, 이후의 프로젝트를 진행해 나가면서 이때의 리뷰를 생각하도록 하겠습니다!! 감사합니다!

wheejuni commented on src/main/java/com/kakao/cafe/web/SessionConst.java:
```diff
@@ -0,0 +1,5 @@
+package com.kakao.cafe.web;
+
+public interface SessionConst {
+    public static final String SESSIONED_USER = "SESSIONED_USER";
```
인터페이스로 상수를 정의하는 것은 좋은 방법이 아닐 겁니다.
인터페이스는 타입이나 스펙을 정의하는 것에만 사용되어야 합니다. 
https://dev.to/kylec32/effective-java-design-user-interface-only-to-define-types-3271

wheejuni replied:
`class` 를 작성하고 생성자를 적절하게 가리는 것만으로도 같은 기능을 수행할 수 있습니다.

```java
public class SessionConstants {
    private SessionConstants () {}

    ...

zbqmgldjfh replied:
좋은 지적 감사드립니다!! 생각해보니 생성자 자체를 막아도 되겠군요!!
반영하도록 하겠습니다!!

wheejuni reviewed (changes requested):
수고 많으셨습니다. 👍 
몇 가지 수정할 점들을 제안해 드립니다~

wheejuni commented:
~질문은 별도 댓글로 답변 드릴게요~ 여기에 바로 답니다.
둘 다 나쁜 방식이 아닌 것 같습니다. 다만 1번의 경우 너무 특정 뷰 프레임워크에 묶인다는 아쉬움은 지우기 어렵겠네요.
그런데 지금 요구사항으론 1번도 충분히 좋습니다~ 취향껏 고고하시죠

zbqmgldjfh committed:
fix(SessionConst): 인터페이스 상수 수정하기

- 수정 전
이전에는 interface 키워드를 사용하여 상수를 만들고 있었다.

- 수정 후
private 생성자를 통해 생성을 방지하고, final 키워드 사용

wheejuni commented:
충분한 대화와 개발이 진행된 것 같아 이제 머지하겠습니다.
